src/ctap2/commands/get_assertion.rs

```diff
@@ -0,0 +1,681 @@
+                .as_ref(),
+        );
+        auth_data.extend_from_slice(self.rp.hash().as_ref());
+        auth_data.extend_from_slice(&[key_handle.len() as u8]);
+        auth_data.extend_from_slice(key_handle.as_ref());
+
+        let cmd = U2F_AUTHENTICATE;
+        let apdu = U2FAPDUHeader::serialize(cmd, flags, &auth_data)?;
+        Ok(apdu)
+    }
+
+    fn handle_response_ctap1(
+        &self,
+        status: Result<(), ApduErrorStatus>,
+        input: &[u8],
+    ) -> Result<Self::Output, Retryable<HIDError>> {
+        if Err(ApduErrorStatus::ConditionsNotSatisfied) == status {
+            return Err(Retryable::Retry);
+        }
+        if status.is_err() {
+            return Err(Retryable::Error(HIDError::DeviceError));
+        }
+
+        named!(
+            parse_authentication<(u8, u32)>,
+            do_parse!(user_presence: be_u8 >> counter: be_u32 >> (user_presence, counter))
+        );
+
+        let (user_presence, counter, signature) = match parse_authentication(input) {
+            Ok((input, (user_presence, counter))) => {
+                let signature = Vec::from(input);
+                Ok((user_presence, counter, signature))
+            }
+            Err(e) => {
+                error!("error while parsing authentication: {:?}", e);
+                Err(io::Error::new(
+                    io::ErrorKind::Other,
+                    "unable to parse authentication",
+                ))
+                .map_err(|e| HIDError::IO(None, e))
+                .map_err(Retryable::Error)
+            }
+        }?;
+
+        let mut flags = AuthenticatorDataFlags::empty();
+        if user_presence == 1 {
+            flags |= AuthenticatorDataFlags::USER_PRESENT;
+        }
+        let auth_data = AuthenticatorData {
+            rp_id_hash: self.rp.hash(),
+            flags,
+            counter,
+            credential_data: None,
+            extensions: Vec::new(),
+        };
+        let assertion = Assertion {
+            credentials: None,
+            signature,
+            public_key: None,
+            auth_data,
+        };
+
+        Ok(AssertionObject(vec![assertion]))
+    }
+}
+
+impl RequestCtap2 for GetAssertion {
+    type Output = AssertionObject;
+
+    fn command() -> Command {
+        Command::GetAssertion
+    }
+
+    fn wire_format<Dev>(&self, _dev: &mut Dev) -> Result<Vec<u8>, HIDError>
+    where
+        Dev: FidoDevice + io::Read + io::Write + fmt::Debug,
+    {
+        // TODO(MS): Add GetInfo-request here and others (See CommandDevice::new)
+        Ok(ser::to_vec(&self).map_err(CommandError::Serialization)?)
+    }
+
+    fn handle_response_ctap2<Dev>(
+        &self,
+        dev: &mut Dev,
+        input: &[u8],
+    ) -> Result<Self::Output, HIDError>
+    where
+        Dev: FidoDevice + io::Read + io::Write + fmt::Debug,
+    {
+        if input.is_empty() {
+            return Err(CommandError::InputTooSmall).map_err(HIDError::Command);
```

Comment:
Same as here as elsewhere, why the redundant map_err?

I won't comment on all instances to not be spammy, but this kind of redundant map_err calls occur in other places in this and other files.